### PR TITLE
Instrumenting granular metrics to understand the RCA quickly

### PR DESCRIPTION
**Pull request by raghumdani**
This commit adds low level metrics to deltacat. Below are the metrics covered as part of this PR:
- Download (latency, failure, success)
- Upload (latency, failure, success)
- Hash bucket (latency, failure, success)
- Merge (latency, failure, success)
- Prepare Deletes (latency, failure, success)
- Discover Deltas (latency, failure, success)
- Materialize (latency, failure, success)

Observed an overhead of 0.02s when actor is not yet available on the node and 0.009s when actor is already available.

### CODE REVIEW

File: deltacat/aws/constants.py [hunk 1]
```diff
@@ -3,6 +3,10 @@
 from deltacat.utils.common import env_integer, env_string
 
 DAFT_MAX_S3_CONNECTIONS_PER_FILE = env_integer("DAFT_MAX_S3_CONNECTIONS_PER_FILE", 8)
-BOTO_MAX_RETRIES = env_integer("BOTO_MAX_RETRIES", 15)
+BOTO_MAX_RETRIES = env_integer("BOTO_MAX_RETRIES", 5)
```
Comment (rkenmi) [inline deltacat/aws/constants.py:6]: Any reason for lowering the defaults if they are adjustable through environment values?

File: deltacat/aws/constants.py [hunk 2]
```diff
@@ -3,6 +3,10 @@
...
 TIMEOUT_ERROR_CODES: List[str] = ["ReadTimeoutError", "ConnectTimeoutError"]
 AWS_REGION = env_string("AWS_REGION", "us-east-1")
+
+# Metric Names
+DOWNLOAD_MANIFEST_ENTRY_METRIC_PREFIX = "download_manifest_entry"
+UPLOAD_SLICED_TABLE_METRIC_PREFIX = "upload_local_table"
```
Comment (rkenmi) [inline deltacat/aws/constants.py:12]: Minor: Should we change this to `upload_sliced_table`?

File: deltacat/utils/metrics.py [hunk 3]
```diff
@@ -177,25 +173,162 @@ def _emit_cloudwatch_emf_metrics(
...
+    metrics_config: MetricsConfig = None
+
+    def set_metrics_config(self, config: MetricsConfig) -> None:
+        self.metrics_config = config
+
+    def get_metrics_config(self) -> MetricsConfig:
+        return self.metrics_config
+
+
+def _emit_with_exception(name: Optional[str], value: Any):
```
Comment (rkenmi) [inline deltacat/utils/metrics.py:201]: Is this private method more like a `_emit_ignore_exceptions`?

**Review by rkenmi — COMMENTED** [review]
LGTM overall. Thanks for the change, this will be helpful.

File: deltacat/aws/constants.py (hunk 1 above)
Comment (raghumdani) [reply deltacat/aws/constants.py:6]: This was discussed during daft offsite. We anyway have tenacity retries.

File: deltacat/aws/constants.py (hunk 2 above)
Comment (raghumdani) [reply deltacat/aws/constants.py:12]: I thought this is meaningful for someone looking at metrics on dashboard.

**Review by rkenmi — APPROVED** [review]

**Pull request merged**